### PR TITLE
TypeInformationDetection-static

### PR DESCRIPTION
**Pull request by arkadius**
## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

### CODE REVIEW

File: engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/typeinformation/TypeInformationDetection.scala [hunk 1]
```diff
@@ -27,3 +32,37 @@ trait TypeInformationDetection extends Serializable {
...
+    ServiceLoader
+      .load(classOf[TypeInformationDetection], classloader)
+      .asScala
+      .toList match {
+      case only :: Nil => only
+      case Nil =>
+        throw new IllegalStateException(
+          s"Missing ${classOf[TypeInformationDetection].getSimpleName} implementation on the classpath. " +
+            s"Classloader: ${printClassloaderDebugDetails(classloader)}. " +
+            s"Ensure that your classpath is correctly configured, probable flinkExecutor.jar is missing"
```
Comment (raphaelsolarski) [inline engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/typeinformation/TypeInformationDetection.scala:51]: `flinkExecutor.jar is probably missing`

File: engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/typeinformation/TypeInformationDetection.scala [hunk 2]
```diff
@@ -27,3 +32,37 @@ trait TypeInformationDetection extends Serializable {
   def forType[T](typingResult: TypingResult): TypeInformation[T]
 
 }
+
+object TypeInformationDetection {
+
+  // We could just inline implementation of TypeInformationDetection but we don't want expose too much implementation
```
Comment (raphaelsolarski) [inline engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/typeinformation/TypeInformationDetection.scala:38]: I don't understand the reason why we use SPI here - imo we should allow it only if we expect that it should be "overridable" in specific deployments.
What you mean by exposing too much implementation?

**Review by raphaelsolarski — CHANGES_REQUESTED** [review]

File: engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/typeinformation/TypeInformationDetection.scala (hunk 2 above)
Comment (raphaelsolarski) [reply engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/typeinformation/TypeInformationDetection.scala:38]: + and we don't want **to** expose too much implementation

Comment (arkadius) [reply engine/flink/components-api/src/main/scala/pl/touk/nussknacker/engine/flink/api/typeinformation/TypeInformationDetection.scala:38]: I was looking for a clean solution that would split single implementation from the interface. I don't want the developer of the component to see too many classes. Do you have some other nice solution that would fulfil this requirement?

**Review by raphaelsolarski — APPROVED** [review]

**Pull request merged**